### DISCUSSION TITLE
Exporting built-in Firedrake mesh to gmsh

### DISCUSSION THREAD

**Discussion by sarshia** [firedrakeproject/firedrake#3399] [Firedrake support]
Hello, is there a way to export a built-in Firedrake mesh to gmsh? For example, if I create a UnitCubeMesh in Firedrake and want to export its node sets, element sets, etc. all to a .msh file for visualization in gmsh. I am able to write the mesh to .pvd but not to .msh.

**Comment by connorjward** [comment]
Unfortunately I do not think that this is possible. Internally we rely on [PETSc DMPlex](https://petsc.org/main/manualpages/DMPlex/) for our mesh representation, and DMPlex only has routines for reading to gmsh files, not writing to them.

**Comment by sarshia** [comment]
Got it, and just to clarify does this include even after we write the mesh to .pvd? In other words, do you know if the node/element data is retained in the .pvd file in a way that could then be exported to gmsh?

**Reply by connorjward** [reply]
At that point you are outside the Firedrake ecosystem so I'm not sure. The Python package [meshio](https://pypi.org/project/meshio/) appears to support loading and writing VTK and .msh files though so might do what you want. I recall this being used by some Firedrake users in the past.
